src/fhir/api.ts

```diff
@@ -1,31 +1,70 @@
+import {FormValues} from "../components/reports/ReportForm";
+import {
+  furtherTestingAndId,
+  organisationAndId,
+  patientAndId, planDefinitionAndId,
+  practitionersAndIds, reportAndId, serviceRequestAndId,
+  specimenAndId,
+  variantAndId,
+} from "./resources";
 
 /**
  * Create a report bundle
- * @param patient FHIR patient entity
+ * @param form values from the front end form
  */
-export const bundleRequest = (patient: typeof patientSchema) => {
+export const bundleRequest = (form: FormValues) => {
   return {
     url: "/",
     method: "POST",
     headers: {'Content-Type': 'application/fhir+json;charset=UTF-8'},
-    body: JSON.stringify(createBundle(patient)),
+    body: JSON.stringify(createBundle(form)),
   };
 }
 
-export const createBundle = (patient: typeof patientSchema) => {
-  const mrn = patient?.identifier?.at(0)?.value;
-  const patientEntry = createPatientEntry(patient);
-
+export const createBundle = (form: FormValues) => {
+  const org = organisationAndId(form.address);
+  const patient = patientAndId(form.patient, org.id);
+  const specimen = specimenAndId(form.sample, patient.id);
+  const furtherTesting = furtherTestingAndId(form.result, patient.id);
+  const plan = planDefinitionAndId(form.sample, form.result, patient.id);
+  const {authoriser, reporter} = practitionersAndIds(form.result);
+  const variant = variantAndId(form.variant, patient.id, specimen.id, specimen.identifier, reporter.id, authoriser.id);
+  const serviceRequest = serviceRequestAndId(form.sample, patient.id, plan.id, reporter.id, specimen.id);
+  const report = reportAndId(form.result, patient.id, reporter.id, authoriser.id, org.id, specimen.id, [variant.id]);
   return {
     resourceType: "Bundle",
     type: "transaction",
-    entry: [{
-      resource: patientEntry,
-      resourceType: "Patient",
-      request: {method: "PUT", url: `Patient?identifier=${mrn}`}
-    }]
+    entry: [
+      createEntry(patient.resource, form.patient.mrn),
+      createEntry(org.resource, org.identifier),
+      createEntry(specimen.resource, specimen.identifier),
+      createEntry(authoriser.resource),
+      createEntry(reporter.resource),
+      createEntry(variant.resource, variant.identifier),
+      createEntry(furtherTesting.resource),
+      createEntry(plan.resource),
+      createEntry(serviceRequest.resource),
+      createEntry(report.resource),
+    ]
   };
 }
 
+/**
+ * Create individual entry for a bundle.
+ *
+ * If identifier is given then the method will be update or create, otherwise only create.
+ * @param resource resource to send
+ * @param identifier identifier used to query on (update if exists or create if not)
+ */
+const createEntry = (resource: Resource, identifier?: string) => {
+  let requestInfo = {method: "POST", url: resource.resourceType}
+  if (identifier !== undefined) {
+    requestInfo = {method: "PUT", url: `${resource.resourceType}?identifier=${identifier}`}
```

Comment:
Yep only these two, the PUT is used to update an existing entity if it exists, otherwise create. the POST is used to always create. This does mean that we should do some work on our end to make sure that a duplicate report is handled correctly